### PR TITLE
Rewrite of CLI using clap

### PR DESCRIPTION
**Pull request by Dherse**
This proposed change, replaces the use of `pico-args` to `clap`. It provides a nicer CLI interface that is no longer handwritten but generated by clap.

# Goal

The goal is to make the CLI code more readable, more maintainable and most importantly easier to expand in the future. Some work is being done to add additional arguments such as #463 . In my (albeit limited) testing, it does not change compiles times significantly (since the LTO linking takes the longest either way) and does not bloat the executable too much, but I look forward to hearing your feedback on these issues.

# Considerations

I have tried using one other CLI argument library with little success, there are lighter alternative to clap, as was discussed on discord [here](https://discord.com/channels/1054443721975922748/1091038400523534417), but right now, clap is still the most popular in the rust ecosystem and has the widest feature set. While that feature set is not fully utilized at the moment, it may be in the future.

For example, I think that adding a verbosity flag and the ability to print log messages in the document would be great, especially printing warnings. These may be best handle with a `-v` flag for setting the verbosity of the log messages.

# Breaking changes

It changes the way subcommands are handled. Rather than handling them using `--fonts` for example, it now uses cargo-like commands:

- `typst fonts` prints the list of fonts
- `typst build` builds the project
- `typst watch` watches the project

Arguments that are command to all commands have been moved to before the subcommand, meaning that `font-path` and `root` must precede the command, here is an example:
`typst --root ../my_project build main.typ`

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -24,145 +23,177 @@ use typst::syntax::{Source, SourceId};
...
-  typst --fonts [OPTIONS]
-
-OPTIONS:
-  -h, --help        Print this help
-  --font-path <dir> Add additional directories to search for fonts
-  --variants        Also list style variants of each font family
-";
+impl CompileSettings {
+    /// Create a new compile settings from the field values.
+    pub fn new(input: PathBuf, output: Option<PathBuf>, watch: bool, root: Option<PathBuf>, font_paths: Vec<PathBuf>) -> Self{
```
Comment (laurmaedje) [inline cli/src/main.rs:112]: could you run cargo fmt?

**Review by laurmaedje — COMMENTED** [review]
Looks good overall

File: cli/src/main.rs [hunk 2]
```diff
@@ -24,145 +23,177 @@ use typst::syntax::{Source, SourceId};
...
+        let output = match output {
+            Some(path) => path,
+            None => {
+                let mut out = input.clone();
```
[the remark this answers is not in the document]
Comment (Dherse) [reply cli/src/main.rs:116]: Yes, I knew you would ask that! setting the extension keeps the rest of the pathbuf intact, I think this is especially relevant if its a complex file path

File: cli/src/main.rs (hunk 1 above)
Comment (Dherse) [reply cli/src/main.rs]: There we go, it's updated

File: cli/src/main.rs [hunk 3]
```diff
@@ -24,145 +23,177 @@ use typst::syntax::{Source, SourceId};
...
+        Self::new(input, output, watch, args.root, args.font_paths)
     }
-}
 
-/// Parse command line arguments.
-fn parse_args() -> StrResult<Command> {
-    let mut args = Arguments::from_env();
-    if args.contains(["-V", "--version"]) {
-        print_version();
+    /// Compile the input file into the output file.
```
[the remark this answers is not in the document]
Comment (Dherse) [reply cli/src/main.rs:144]: I fully agree, I just thought it looked nicer, it really isn't much of a change to do, in general I just prefer to do `my_struct.this()` on a struct than `this(my_struct)`, I can change it if you want.

[the remark this answers is not in the document]
Comment (laurmaedje) [reply cli/src/main.rs:144]: I agree that it often looks nicer, but in this case it doesn't pull its weight imo

File: cli/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (laurmaedje) [reply cli/src/main.rs:122]: Doesn't `with_extension` do the same thing? https://doc.rust-lang.org/src/std/path.rs.html#2604

[the remark this answers is not in the document]
Comment (Dherse) [reply cli/src/main.rs:122]: Yep, changing it, I don't know why I thought it did something else. My bad

File: cli/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dherse) [reply cli/src/main.rs]: Fixed in latest commits

**Comment by laurmaedje** [comment]
Ready for merge?

**Comment by Dherse** [comment]
For me it's all good :D

**Comment by laurmaedje** [comment]
Thanks!

**Comment by Dherse** [comment]
> The most breaking change is `typst file.typ` does not work anymore...

That is true, it has been discussed on Discord with @laurmaedje and it was agreed that as the CLI tool becomes more complex, especially once package management features are included, it will no longer be reasonable to offer this default command. I will add it to the top level to make it clearer :)

**Pull request merged**